logger.go

```diff
@@ -96,3 +112,72 @@ func (l *MatchingLogger) Write(p []byte) (int, error) {
 
 	return len(p), nil
 }
+
+// parsedLog is a lightweight log structure we parse from subsystem logs.
+// Another approach for capturing logs from subsystems is to pass around
+// LogEntry or opentelemetry structs.
+type parsedLog struct {
+	date       []byte
+	level      []byte
+	loggerName []byte
+	location   []byte
+	tail       []byte
+}
+
+func (p parsedLog) valid() bool {
+	return len(p.date) > 0 && len(p.level) > 0 && len(p.loggerName) > 0 && len(p.location) > 0 && len(p.tail) > 0
+}
+
+var levels = map[string]zapcore.Level{
+	"DEBUG":  zapcore.DebugLevel,
+	"INFO":   zapcore.InfoLevel,
+	"WARN":   zapcore.WarnLevel,
+	"ERROR":  zapcore.ErrorLevel,
+	"DPANIC": zapcore.DPanicLevel,
+	"PANIC":  zapcore.PanicLevel,
+	"FATAL":  zapcore.FatalLevel,
+}
+
+var colorCodeRegexp = regexp.MustCompile(`\x1b\[\d+m`)
+
+// stripAnsiColorCodes removes color codes from a string so we can use it internally.
+func stripAnsiColorCodes(raw []byte) []byte {
+	return colorCodeRegexp.ReplaceAll(raw, nil)
+}
+
+// entry converts a parsedLog to a zapcore.Entry which can be NetAppender'd.
+func (p parsedLog) entry() zapcore.Entry {
+	level, ok := levels[string(p.level)]
+	if !ok {
+		level = zapcore.DebugLevel
+	}
+	file, rawLine, defined := bytes.Cut(p.location, []byte{':'})
+	line, _ := strconv.ParseUint(string(rawLine), 10, 64) //nolint:errcheck
+	return zapcore.Entry{
+		Level: level,
+		// note: time.Now() is basically correct, and simpler than parsing.
+		Time:       time.Now().UTC(),
+		LoggerName: string(p.loggerName),
+		Message:    string(p.tail),
+		Caller:     zapcore.EntryCaller{Defined: defined, File: string(file), Line: int(line)},
+	}
+}
+
+// getIndexOrNil returns the element at `index` of `arr`, or nil if out of range.
+func getIndexOrNil[T any](arr [][]T, index int) []T {
+	if index < len(arr) {
+		return arr[index]
+	}
+	return nil
+}
+
+func parseLog(line []byte) *parsedLog {
```

Comment:
I feel this needs to return an error instead of using the getIndexOrNil() stuff. If we can't parse the line, I think it's important to consider that an error and write the WHOLE line to the log. That is, revert to the "Unstructured Log Output" behavior that previously existed.) Otherwise, a parsing failure means we could lose critical information when it's likely the most critical to know it (e.g. something unexpected is happening.)

Reply:
done in eb094112 + added tests to make sure the parser fallback behaves as expected